presto-tests/src/main/java/com/facebook/presto/tests/DistributedQueryRunner.java

```diff
@@ -136,19 +139,42 @@ private DistributedQueryRunner(
             discoveryServer = new TestingDiscoveryServer(environment);
             closer.register(() -> closeUnchecked(discoveryServer));
             log.info("Created TestingDiscoveryServer in %s", nanosSince(start).convertToMostSuccinctTimeUnit());
+            URI discoveryUrl = discoveryServer.getBaseUrl();
+            log.info("Discovery URL %s", discoveryUrl);
+
+            if (nativeWorkerLauncher.isPresent()) {
+                ImmutableList.Builder<Process> nativeWorkersBuilder = ImmutableList.builder();
+                for (int i = 0; i < nodeCount; i++) {
+                    nativeWorkersBuilder.add(nativeWorkerLauncher.get().apply(i, discoveryUrl));
+                }
+                nativeWorkers = nativeWorkersBuilder.build();
+                closer.register(() -> {
+                    for (Process nativeWorker : nativeWorkers) {
+                        nativeWorker.destroyForcibly();
+                    }
+                });
+            }
+            else {
+                nativeWorkers = ImmutableList.of();
+            }
 
             ImmutableList.Builder<TestingPrestoServer> servers = ImmutableList.builder();
-
-            for (int i = 1; i < nodeCount; i++) {
-                TestingPrestoServer worker = closer.register(createTestingPrestoServer(discoveryServer.getBaseUrl(), false, extraProperties, parserOptions, environment, baseDataDir));
-                servers.add(worker);
+            if (!nativeWorkerLauncher.isPresent()) {
```

Comment:
It feels like we can fairly easily squash these three checks on `nativeWorkerLauncher` into one?